### ISSUE TITLE
previous node taint is wiped when kubelet restarts

### ISSUE THREAD

**Issue by Chaunceyctx** [kubernetes/kubernetes#119645]
### What happened?

I have a k8s cluster(v1.27.2) containing one node. I set `evictionHard: nodefs.available: 90%` and write a large amount of data to kubelet rootdir（used 8GB / total 10GB ） to trigger eviction. `node.kubernetes.io/disk-pressure` taint was added to this node. But when kubelet restarted, previous disk-pressure taint was wiped weirdly. And pending pod is normally scheduled to run on the current node. Then I checked the kubelet logs and found:

kubelet restart:
```
I0727 11:09:05.782733 3813426 flags.go:64] FLAG: --tls-private-key-file=""
I0727 11:09:05.782737 3813426 flags.go:64] FLAG: --topology-manager-policy="none"
I0727 11:09:05.782741 3813426 flags.go:64] FLAG: --topology-manager-scope="container"
I0727 11:09:05.782745 3813426 flags.go:64] FLAG: --v="8"
I0727 11:09:05.782749 3813426 flags.go:64] FLAG: --version="false"
I0727 11:09:05.782754 3813426 flags.go:64] FLAG: --vmodule=""
I0727 11:09:05.782757 3813426 flags.go:64] FLAG: --volume-plugin-dir="/usr/libexec/kubernetes/kubelet-plugins/volume/exec/"
I0727 11:09:05.782761 3813426 flags.go:64] FLAG: --volume-stats-agg-period="1m0s"
I0727 11:09:05.796690 3813426 mount_linux.go:222] Detected OS with systemd
```
update node status **NodeHasNoDiskPressure**
``` 
I0727 11:09:10.858628 3813426 kubelet_node_status.go:632] "Recording event message for node" node="192.168.2.107" event="NodeHasNoDiskPressure"
I0727 11:09:10.858640 3813426 kubelet_node_status.go:762] "Setting node status condition code" position=9 node="192.168.2.107"
```
eviction manager  start to synchronize 
```
I0727 11:09:13.373393 3813426 eviction_manager.go:292] "Eviction manager: synchronize housekeeping"
```

Q: Why does kubelet report **NodeHasNoDiskPressure** ?
A: Eviction manager has not yet executed the synchronize method

Normally, a node that was previously marked with taints should retain after reboot otherwise it doesn't make sense. Maybe we need to make sure that taints can't be deleted when kubelet restarts?

### What did you expect to happen?

previous node taint is not wiped 

### How can we reproduce it (as minimally and precisely as possible)?

Restart kubelet repeatedly after the disk pressure eviction is triggered. Observe `node.spec.taints`

### Kubernetes version

1.27.2

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by Chaunceyctx** [comment]
@bobbypage @liggitt  PTAL

**Comment by Chaunceyctx** [comment]
@klueska @dashpole @liggitt PTAL

**Comment by heartlock** [comment]
I have encountered the same bug.

**Comment by SuQiucheng** [comment]
I hope the community can fix this bug as soon as possible.

**Comment by waiterQ** [comment]
I also met this before, is there a fix plan?

**Comment by Chaunceyctx** [comment]
@ffromani @tzneal I believe that the impact of this problem is actually very small, and the taints of the node will disappear briefly and then appear again. Therefore, I would like to ask if it is still necessary for us to repair this problem

**Comment by SuQiucheng** [comment]
I also meet this problem, I wonder if there's any restoration value

**Comment by pacoxu** [comment]
> @ffromani @tzneal I believe that the impact of this problem is actually very small, and the taints of the node will disappear briefly and then appear again. Therefore, I would like to ask if it is still necessary for us to repair this problem

I feel this is a bug that should be fixed, although the impact doesn't seem particularly significant. 

During the kubelet startup process, a check should be performed when setting the condition, rather than clearing the taints first.

cc @SergeyKanzhelev @kannon92

**Comment by Chaunceyctx** [comment]
> > @ffromani @tzneal I believe that the impact of this problem is actually very small, and the taints of the node will disappear briefly and then appear again. Therefore, I would like to ask if it is still necessary for us to repair this problem
> 
> I feel this is a bug that should be fixed, although the impact doesn't seem particularly significant.
> 
> During the kubelet startup process, a check should be performed when setting the condition, rather than clearing the taints first.
> 
> cc @SergeyKanzhelev @kannon92

yes, I think this is a TOCTOU bug. maybe I will rethink how to fix it. previous pr cannot fix this issue.

**Comment by pacoxu** [comment]
@Chaunceyctx can you reproduce it with v1.32? I cannot reproduce it with v1.32 locally. 

I tried v1.27 and I can reproduce it. I am not sure why. Can you give it a try?

**Comment by Chaunceyctx** [comment]
> @Chaunceyctx can you reproduce it with v1.32? I cannot reproduce it with v1.32 locally.
> 
> I tried v1.27 and I can reproduce it. I am not sure why. Can you give it a try?

ok, I will try to reproduce it with v1.32

**Comment by Chaunceyctx** [comment]
@pacoxu  I have tested it on version 1.32, and it is indeed quite hard to reproduce. However, if I introduce a delay of about 1 second to `m.summaryProvider.Get` to simulate a scenario with a large number of containers on the node(`m.summaryProvider.Get` method will takes some time to get container metrics recursively), 
https://github.com/kubernetes/kubernetes/blob/a32f69590a9585adb12952dbc13f7137f37f49bf/pkg/kubelet/eviction/eviction_manager.go#L290
```
summary, err := m.summaryProvider.Get(ctx, updateStats)
time.Sleep(time.Second)
```


this issue reappears.

**Comment by haircommander** [comment]
/retitle previous node condition (disk pressure) is wiped when kubelet restarts
